src/pypdfium2/_helpers/document.py

```diff
@@ -72,6 +73,8 @@ def __init__(self, input, password=None, autoclose=False, to_close=None):
         
         # CONSIDER adding a public method to inject into data_holder/data_closer
         
+        # NOTE due to mmap + weakref.finalize idiosyncrasies, we cannot hold a reference to the actual input on class level, otherwise we run into "cannot close; exported pointers exist" trouble
```

Comment:
Theoretically we could if we were able to `del self._input` (or `self._input = None`) on before closing, but we cannot access `self` in the finalizer callback, so we cannot get rid of the reference.
weakref is awkward here: we cannot access the object that is being finalized, but apparently it still exists at that time.